File: test/correctness/multipass_constraints.cpp
#include "Halide.h"
#include <gtest/gtest.h>

using namespace Halide;

TEST(MultipassConstraintsTest, MultipassConstraints) {
    ImageParam in(Float(32), 2, "in");

    Func out("out");
    Var x("x"), y("y");

    out(x, y) = in(x + 1, y + 1) + in(x - 1, y - 1);
    out(x, y) += 3.0f;
    out.update().vectorize(x, 4);

    OutputImageParam o = out.output_buffer();

    // Now make some hard-to-resolve constraints
    in.dim(0).set_bounds(in.dim(1).min() - 5, in.dim(1).extent() + o.dim(0).extent());

    o.dim(0).set_bounds(0, select(o.dim(0).extent() < 22, o.dim(0).extent() + 1, o.dim(0).extent()));

    // Make a bounds query buffer
    Buffer<float> out_buf(nullptr, 7, 8);
    out_buf.set_min(2, 2);

    ASSERT_NO_THROW(out.infer_input_bounds(out_buf));

    EXPECT_EQ(in.get().dim(0).min(), -4);
    EXPECT_EQ(in.get().dim(0).extent(), 34);
    EXPECT_EQ(in.get().dim(1).min(), 1);
    EXPECT_EQ(in.get().dim(1).extent(), 10);
    EXPECT_EQ(out_buf.dim(0).min(), 0);
    EXPECT_EQ(out_buf.dim(0).extent(), 24);
    EXPECT_EQ(out_buf.dim(1).min(), 2);
    EXPECT_EQ(out_buf.dim(1).extent(), 8);
}

File: python_bindings/test/correctness/multipass_constraints.py
import halide as hl


def test_multipass_constraints():
    input = hl.ImageParam(hl.Float(32), 2, "input")

    f = hl.Func("f")
    x = hl.Var("x")
    y = hl.Var("y")

    f[x, y] = input[x + 1, y + 1] + input[x - 1, y - 1]
    f[x, y] += 3.0
    f.update().vectorize(x, 4)

    o = f.output_buffer()

    # Now make some hard-to-resolve constraints
    input.dim(0).set_bounds(
        min=input.dim(1).min() - 5, extent=input.dim(1).extent() + o.dim(0).extent()
    )

    o.dim(0).set_bounds(
        min=0,
        extent=hl.select(
            o.dim(0).extent() < 22, o.dim(0).extent() + 1, o.dim(0).extent()
        ),
    )

    # Make a bounds query buffer
    query_buf = hl.Buffer.make_bounds_query(type=hl.Float(32), sizes=[7, 8])
    query_buf.set_min([2, 2])

    f.infer_input_bounds(query_buf)

    if (
        input.get().dim(0).min() != -4
        or input.get().dim(0).extent() != 34
        or input.get().dim(1).min() != 1
        or input.get().dim(1).extent() != 10
        or query_buf.dim(0).min() != 0
        or query_buf.dim(0).extent() != 24
        or query_buf.dim(1).min() != 2
        or query_buf.dim(1).extent() != 8
    ):

        print(
            "Constraints not correctly satisfied:\n",
            "in:",
            input.get().dim(0).min(),
            input.get().dim(0).extent(),
            input.get().dim(1).min(),
            input.get().dim(1).extent(),
            "out:",
            query_buf.dim(0).min(),
            query_buf.dim(0).extent(),
            query_buf.dim(1).min(),
            query_buf.dim(1).extent(),
        )
        assert False


if __name__ == "__main__":
    test_multipass_constraints()
